src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ActionPlanSessionsService.kt

```diff
@@ -0,0 +1,192 @@
+
+    if (appointment.appointmentFeedbackSubmittedAt != null) {
+      throw ResponseStatusException(HttpStatus.CONFLICT, "session feedback has already been submitted for this session")
+    }
+
+    setBehaviourFields(appointment, behaviourDescription, notifyProbationPractitioner)
+    appointmentRepository.save(appointment)
+    return actionPlanSessionRepository.save(session)
+  }
+
+  fun submitSessionFeedback(actionPlanId: UUID, sessionNumber: Int, submitter: AuthUser): ActionPlanSession {
+    val session = getActionPlanSessionOrThrowException(actionPlanId, sessionNumber)
+    val appointment = session.currentAppointment
+
+    if (appointment?.appointmentFeedbackSubmittedAt != null) {
+      throw ResponseStatusException(HttpStatus.CONFLICT, "session feedback has already been submitted for this session")
+    }
+
+    if (appointment?.attendanceSubmittedAt == null) {
+      throw ResponseStatusException(HttpStatus.UNPROCESSABLE_ENTITY, "can't submit session feedback unless attendance has been recorded")
+    }
+
+    appointment.appointmentFeedbackSubmittedAt = OffsetDateTime.now()
+    appointment.appointmentFeedbackSubmittedBy = submitter
+    actionPlanSessionRepository.save(session)
+
+    appointmentEventPublisher.attendanceRecordedEvent(session, appointment.attended!! == Attended.NO)
+    appointmentEventPublisher.behaviourRecordedEvent(session, appointment.notifyPPOfAttendanceBehaviour!!)
+    appointmentEventPublisher.sessionFeedbackRecordedEvent(session, appointment.notifyPPOfAttendanceBehaviour!!)
+    return session
+  }
+
+  fun getSessions(actionPlanId: UUID): List<ActionPlanSession> {
+    return actionPlanSessionRepository.findAllByActionPlanId(actionPlanId)
+  }
+
+  fun getSession(actionPlanId: UUID, sessionNumber: Int): ActionPlanSession {
+    return getActionPlanSessionOrThrowException(actionPlanId, sessionNumber)
+  }
+
+  private fun setAttendanceFields(
+    appointment: Appointment,
+    attended: Attended,
+    additionalInformation: String?
+  ) {
+    appointment.attended = attended
+    additionalInformation?.let { appointment.additionalAttendanceInformation = additionalInformation }
+    appointment.attendanceSubmittedAt = OffsetDateTime.now()
+  }
+
+  private fun setBehaviourFields(
+    appointment: Appointment,
+    behaviour: String,
+    notifyProbationPractitioner: Boolean,
+  ) {
+    appointment.attendanceBehaviour = behaviour
+    appointment.attendanceBehaviourSubmittedAt = OffsetDateTime.now()
+    appointment.notifyPPOfAttendanceBehaviour = notifyProbationPractitioner
+  }
+
+  private fun checkSessionIsNotDuplicate(actionPlanId: UUID, sessionNumber: Int) {
+    getActionPlanSession(actionPlanId, sessionNumber)?.let {
+      throw EntityExistsException("Action plan session already exists for [id=$actionPlanId, sessionNumber=$sessionNumber]")
+    }
+  }
```

Comment:
This is not guaranteed to work as it is not atomic - for instance when the use click on the button twice in quick succession.

Perhaps have a constraint on the database where actionPlanId and sessionNumber has to be unique

Reply:
good point - this bug is inherited. shall we fix outside the scope of this PR? would you like to put in a quick PR?

Reply:
or do you think i should just include in these changes if we're just talking about a sql constraint?

Reply:
sorry, I just noticed. I personally think create a separate bug ticket is best - as maybe might want to write some unit tests for it which won't be as quick

Reply:
https://dsdmoj.atlassian.net/browse/IC-1897